src/chain_sync/chain_follower.rs

```diff
@@ -0,0 +1,969 @@
+                            network.clone(),
+                            state_manager.clone(),
+                            bad_block_cache.clone(),
+                            stateless_mode,
+                        );
+                        tokio::spawn({
+                            let state_machine = state_machine.clone();
+                            let state_changed = state_changed.clone();
+                            async move {
+                                if let Some(event) = action.await {
+                                    state_machine.lock().update(event);
+                                    state_changed.notify_one();
+                                }
+                                tasks_clone.lock().remove(&task);
+                            }
+                        });
+                    }
+                }
+            }
+        }
+    });
+
+    // Periodically report progress if there are any tipsets left to be fetched.
+    // Once we're in steady-state (i.e. caught up to HEAD) and there are no
+    // active forks, this will not report anything.
+    set.spawn({
+        let state_manager = state_manager.clone();
+        let state_machine = state_machine.clone();
+        async move {
+            loop {
+                tokio::time::sleep(tokio::time::Duration::from_secs(10)).await;
+                let (tasks_set, _) = state_machine.lock().tasks();
+                let heaviest_epoch = state_manager.chain_store().heaviest_tipset().epoch();
+                let fork_cutoff = heaviest_epoch
+                    - SECONDS_IN_DAY / (state_manager.chain_config().block_delay_secs as i64);
+
+                // Count tipsets to fetch for main chain and forks
+                let mut main_chain_epochs = Vec::new();
+                let mut fork_tipsets = 0;
+
+                for task in tasks_set.iter() {
+                    if let SyncTask::FetchTipset(_, epoch) = task {
+                        let diff = epoch - heaviest_epoch;
+                        if diff >= 0 {
+                            main_chain_epochs.push(diff);
+                        } else {
+                            // This is a fork - we'll download a fixed number of tipsets
+                            fork_tipsets = fork_tipsets.max(epoch - fork_cutoff);
+                        }
+                    }
+                }
+
+                let to_download = main_chain_epochs
+                    .iter()
+                    .max()
+                    .copied()
+                    .unwrap_or_default()
+                    .max(fork_tipsets);
+                let expected_head = calculate_expected_epoch(
+                    Utc::now().timestamp() as u64,
+                    state_manager.chain_store().genesis_block_header().timestamp,
+                    state_manager.chain_config().block_delay_secs,
+                );
+
+                match (expected_head as i64 - heaviest_epoch > 10, to_download > 0) {
```

Comment:
Where does this 10 come from? Let's have it named and potentially configurable.

Reply:
It's an arbitrary number to avoid being too verbose.